sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverter.scala

```diff
@@ -0,0 +1,226 @@
+   * @param baseOffset the base offset of the destination address
+   * @return the number of bytes written. This should be equal to `getSizeRequirement(row)`.
+   */
+  def writeRow(row: Row, baseObject: Object, baseOffset: Long): Long = {
+    unsafeRow.pointTo(baseObject, baseOffset, writers.length, null)
+    var fieldNumber = 0
+    var appendCursor: Int = fixedLengthSize
+    while (fieldNumber < writers.length) {
+      if (row.isNullAt(fieldNumber)) {
+        unsafeRow.setNullAt(fieldNumber)
+      } else {
+        appendCursor += writers(fieldNumber).write(
+          row(fieldNumber),
+          fieldNumber,
+          unsafeRow,
+          baseObject,
+          baseOffset,
+          appendCursor)
+      }
+      fieldNumber += 1
+    }
+    appendCursor
+  }
+
+}
+
+/**
+ * Function for writing a column into an UnsafeRow.
+ */
+private abstract class UnsafeColumnWriter[T] {
+  /**
+   * Write a value into an UnsafeRow.
+   *
+   * @param value the value to write
+   * @param columnNumber what column to write it to
+   * @param row a pointer to the unsafe row
+   * @param baseObject the base object of the target row's address
+   * @param baseOffset the base offset of the target row's address
+   * @param appendCursor the offset from the start of the unsafe row to the end of the row;
+   *                     used for calculating where variable-length data should be written
+   * @return the number of variable-length bytes written
+   */
+  def write(
+      value: T,
+      columnNumber: Int,
+      row: UnsafeRow,
+      baseObject: Object,
+      baseOffset: Long,
+      appendCursor: Int): Int
+
+  /**
+   * Return the number of bytes that are needed to write this variable-length value.
+   */
+  def getSize(value: T): Int
+}
+
+private object UnsafeColumnWriter {
+
+  def forType(dataType: DataType): UnsafeColumnWriter[_] = {
+    dataType match {
+      case IntegerType => IntUnsafeColumnWriter
+      case LongType => LongUnsafeColumnWriter
+      case FloatType => FloatUnsafeColumnWriter
+      case DoubleType => DoubleUnsafeColumnWriter
+      case StringType => StringUnsafeColumnWriter
+      case t =>
+        throw new UnsupportedOperationException(s"Do not know how to write columns of type $t")
+    }
+  }
+}
+
+// ------------------------------------------------------------------------------------------------
+
+private object IntUnsafeColumnWriter extends IntUnsafeColumnWriter
+private object LongUnsafeColumnWriter extends LongUnsafeColumnWriter
+private object FloatUnsafeColumnWriter extends FloatUnsafeColumnWriter
+private object DoubleUnsafeColumnWriter extends DoubleUnsafeColumnWriter
+private object StringUnsafeColumnWriter extends StringUnsafeColumnWriter
+
+private abstract class PrimitiveUnsafeColumnWriter[T] extends UnsafeColumnWriter[T] {
+  def getSize(value: T): Int = 0
+}
+
+private class IntUnsafeColumnWriter private() extends PrimitiveUnsafeColumnWriter[Int] {
+  override def write(
```

Comment:
More specifically, this is the number of variable-length bytes written.  Primitive types, like int, don't write to the variable-length portion and thus return 0.